### PR TITLE
Porting over some low cost Algorithms

### PR DESCRIPTION
**Pull request by avik-pal**
- [x] Broyden
  - [x] Inplace
  - [x] OOP
  - [x] Integrate into the Test Suite
    - [x] Basic Tests
    - [x] 23 Test Problems
- [x] Klement
  - [x] Inplace
  - [x] OOP
  - [x] Integrate into the Test Suite
    - [x] Basic Tests
    - [x] 23 Test Problems

### CODE REVIEW

**Comment by avik-pal** [comment]
For `Klement` if we allow any general linear solve, how do we go about doing the singularity check for resetting? For `LU`, we could do the check on the diagonal of `U`, but that won't work for other solvers. Should we use `cond(J) > inv(tol)`, or is there a better way to do this?

cc @ChrisRackauckas

**Comment by avik-pal** [comment]
I have added both `Klement` and `Broyden`. The naming is a bit unfortunate since `SimpleNonlinearSolve` reserves the names instead of `Simple****`.

Broyden is faster than the one in SimpleNonlinearSolve.

Klement as expected seems to spend a significant amount of time in singularity checking. Someone with more familiarity with LinearAlgebra needs to chime in on how to handle this in the general case. I can add a branching based on LU factorization and default to that.

**Comment by avik-pal** [comment]
I am reusing the factorization for singularity if the user specifies `LUFactorization`, else we fallback to ArrayInterface.issingular

```julia
julia> @benchmark solve(probN, GeneralKlement(; linsolve=LUFactorization()))
BenchmarkTools.Trial: 23 samples with 1 evaluation.
 Range (min … max):  168.920 ms … 283.473 ms  ┊ GC (min … max): 0.00% … 40.56%
 Time  (median):     218.867 ms               ┊ GC (median):    1.57%
 Time  (mean ± σ):   219.950 ms ±  37.673 ms  ┊ GC (mean ± σ):  8.37% ± 11.92%

  ███           ▁ █   ▁   ▁ ▁     ▁   █ ▁ ▁      ▁█ ▁     ▁   ▁  
  ███▁▁▁▁▁▁▁▁▁▁▁█▁█▁▁▁█▁▁▁█▁█▁▁▁▁▁█▁▁▁█▁█▁█▁▁▁▁▁▁██▁█▁▁▁▁▁█▁▁▁█ ▁
  169 ms           Histogram: frequency by time          283 ms <

 Memory estimate: 252.29 MiB, allocs estimate: 201.

julia> @benchmark solve(probN, GeneralKlement(; linsolve=QRFactorization()))
BenchmarkTools.Trial: 12 samples with 1 evaluation.
 Range (min … max):  382.058 ms … 475.142 ms  ┊ GC (min … max): 1.69% … 20.59%
 Time  (median):     440.708 ms               ┊ GC (median):    1.52%
 Time  (mean ± σ):   433.578 ms ±  36.193 ms  ┊ GC (mean ± σ):  5.63% ±  7.69%

  █ ▁           ▁                    ▁ ▁ ▁       ▁ ▁        ▁ █  
  █▁█▁▁▁▁▁▁▁▁▁▁▁█▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█▁█▁█▁▁▁▁▁▁▁█▁█▁▁▁▁▁▁▁▁█▁█ ▁
  382 ms           Histogram: frequency by time          475 ms <

 Memory estimate: 309.65 MiB, allocs estimate: 229.

julia> @benchmark solve(probN, GeneralKlement(; linsolve=nothing))
BenchmarkTools.Trial: 18 samples with 1 evaluation.
 Range (min … max):  232.665 ms … 382.726 ms  ┊ GC (min … max):  0.87% … 36.40%
 Time  (median):     274.173 ms               ┊ GC (median):     2.18%
 Time  (mean ± σ):   283.057 ms ±  47.992 ms  ┊ GC (mean ± σ):  10.11% ± 12.59%

  █▁   ▁                                                         
  ██▁▁▁█▁▁▁▁▁▁▆▁▁▁▁▁▁▁▁▆▁▁▁▁▁▆▁▁▁▁▁▁▆▆▁▆▆▆▁▁▁▆▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▆ ▁
  233 ms           Histogram: frequency by time          383 ms <

 Memory estimate: 305.70 MiB, allocs estimate: 192.
```

on a problem with `1000` states.

**Comment by avik-pal** [comment]
Marking it as WIP, I want to add LBroyden to this as well!

**Comment by avik-pal** [comment]
@ChrisRackauckas this is ready to go!

**Comment by avik-pal** [comment]
@ChrisRackauckas let's merge this one?

File: docs/src/solvers/NonlinearSystemSolvers.md [hunk 1]
```diff
@@ -65,6 +65,10 @@ features, but have a bit of overhead on very small problems.
   - `FastShortcutNonlinearPolyalg()`: The default method. A polyalgorithm that mixes fast methods
     with fallbacks to robust methods to allow for solving easy problems quickly without sacrificing
     robustnes on the hard problems.
+  - `GeneralBroyden()`: Generalization of Broyden's Quasi-Newton Method with Line Search and
+    Automatic Jacobian Resetting. This is a fast method but unstable for most problems!
+  - `GeneralKlement()`: Generalization of Klement's Quasi-Newton Method with Line Search and
+    Automatic Jacobian Resetting. This is a fast method but unstable for most problems!
```
Comment (ChrisRackauckas) [inline docs/src/solvers/NonlinearSystemSolvers.md:71]:
```suggestion
    Automatic Jacobian Resetting. This is a fast method but unstable when the condition number of
    the Jacobian matrix is sufficiently large.
  - `GeneralKlement()`: Generalization of Klement's Quasi-Newton Method with Line Search and
    Automatic Jacobian Resetting. This is a fast method but unstable when the condition number of
    the Jacobian matrix is sufficiently large.
```

**Pull request merged**